### PR TITLE
Motor simulation

### PR DESCRIPTION
**Pull request by KangarooKoala**
Added sim folder with PhysicsSim, SimProfile, and TalonFXSimProfile
Added simulationInit and simulationPeriodic to Robot

### CODE REVIEW

**Comment by KangarooKoala** [comment]
Still need to set reasonable motor parameters in simulation.

**Review by bobtseattle — COMMENTED** [review]
I don't have much experience with physics sim yet, but this seems headed in the right direction.

**Review by TAKBS2412 — APPROVED** [review]
Works in sim

File: src/main/java/frc/team2412/robot/sim/TalonFXSimProfile.java [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+
+public class TalonFXSimProfile extends SimProfile {
+    private final TalonFXSimCollection falconSimCollection;
+    private final double accelToFullTime;
+    private final double fullVel;
+    private final boolean sensorPhase;
+
+    // TODO Log pos????
+    private double pos = 0;
+    private double vel = 0;
```
Comment (jamesdooley4) [inline src/main/java/frc/team2412/robot/sim/TalonFXSimProfile.java:16]: Logging here shouldn't be needed as the existing motor logging in the subsystems (if any) will already log that

File: src/main/java/frc/team2412/robot/Robot.java [hunk 2]
```diff
@@ -191,4 +194,33 @@ public void teleopInit() {
     public void autonomousExit() {
         CommandScheduler.getInstance().cancelAll();
     }
+
+    @Override
+    public void simulationInit() {
+        PhysicsSim physicsSim = PhysicsSim.getInstance();
+        // TODO Find more accurate values
```
Comment (jamesdooley4) [inline src/main/java/frc/team2412/robot/Robot.java:201]: For drive, I've been using 6000 RPM as the top speed.  Seems odd that turret would run slower, as I don't expect there's a lot more torque needed from it.

Comment (jamesdooley4) [reply src/main/java/frc/team2412/robot/Robot.java:201]: And highly recommend adding a comment describing what the numbers mean (RPM, ticks per rotation, conversion from  minutes to 100ms).

File: src/main/java/frc/team2412/robot/Robot.java [hunk 3]
```diff
@@ -191,4 +194,33 @@ public void teleopInit() {
...
+        }
+        if (INDEX_ENABLED) {
+            physicsSim.addTalonFX(hardware.ingestIndexMotor, 1, 5000 * 2048 / 600.);
+            physicsSim.addTalonFX(hardware.feederIndexMotor, 1, 5000 * 2048 / 600.);
+        }
+        if (SHOOTER_ENABLED) {
+            physicsSim.addTalonFX(hardware.flywheelMotor1, 3, 5000 * 2048 / 600.);
+            physicsSim.addTalonFX(hardware.flywheelMotor2, 3, 5000 * 2048 / 600.);
+            physicsSim.addTalonFX(hardware.turretMotor, 2, 4000 * 2048 / 600.);
+            physicsSim.addSparkMax(hardware.hoodMotor, 0.97, 11000); // Incorporate gearing?
```
[the remark this answers is not in the document]
Comment (jamesdooley4) [reply src/main/java/frc/team2412/robot/Robot.java:218]: And how is 11000 calculated?  It'd be good to have the same format as above max velocity constants

**Review by jamesdooley4 — CHANGES_REQUESTED** [review]

File: src/main/java/frc/team2412/robot/sim/SimProfile.java [hunk 4]
```diff
@@ -0,0 +1,29 @@
...
+     *
+     * @return Time since last call, in milliseconds.
+     */
+    protected double getPeriod() {
+        if (!running) {
+            lastTime = System.nanoTime();
+            running = true;
+        }
+
+        long now = System.nanoTime();
```
[the remark this answers is not in the document]
Comment (jamesdooley4) [reply src/main/java/frc/team2412/robot/sim/SimProfile.java:23]: Or use the edu.wpi.first.wpilibj.Timer class (which internally uses RobotController.getFPGATime) and exposes a number of useful helper methods.

[the remark this answers is not in the document]
Comment (KangarooKoala) [reply src/main/java/frc/team2412/robot/sim/SimProfile.java:23]: We only use the times to calculate how much time elapsed, but if you think it'd be helpful to get the times relative to the sim, I could change it. If the simulation clock is pausable, that'd be another reason to switch to FPGA time or the Timer class.

[the remark this answers is not in the document]
Comment (jamesdooley4) [reply src/main/java/frc/team2412/robot/sim/SimProfile.java:23]: Yes, the FPGA time is pausable in simulation.  And the Timer() class itself has built-in support for calculating elapsed time (see the reset() method).

File: src/main/java/frc/team2412/robot/Robot.java (hunk 2 above)
Comment (KangarooKoala) [reply src/main/java/frc/team2412/robot/Robot.java:203]: I switched to using constants, which should clarify what's going on. (The place where the constant is calculated also has a comment explaining it.) I also added a comment on the parameters of PhysicsSim.add*().

**Comment by KangarooKoala** [comment]
I can't reproduce the compile error (It couldn't find the `intakeMotor1` and `intakeMotor2` fields of `hardware`) on my machine, hopefully it doesn't show up again.

**Comment by KangarooKoala** [comment]
> I can't reproduce the compile error (It couldn't find the `intakeMotor1` and `intakeMotor2` fields of `hardware`) on my machine, hopefully it doesn't show up again.

The check failed again for the same reason, even though `intakeMotor1` and `intakeMotor2` are fields of `hardware` and `./gradlew check` works on my machine.

**Review by jamesdooley4 — APPROVED** [review]

**Pull request merged**